File: backend/Bussiness/Services/TourAndTravels/DepartureManagementService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using AutoMapper;
using Domain.Models.TourAndTravels;
using Repository.Interfaces.TourAndTravels;
using static Domain.Models.TourAndTravels.DepartureManagement;
using RepDTO = Repository.DataModels.TourAndTravels.DepartureManagementDTO;

namespace Bussiness.Services.TourAndTravels
{
    // ================================================================
    // Service Interface
    // ================================================================
    public interface IDepartureManagementService
    {
        List<BookingGroupByItinerary> GetUnassignedBookingsGrouped();
        List<SuggestedDepartureItem> GetSuggestedDepartures(long itineraryId, int requiredSeats);
        DepartureDetailResponse CreateDeparture(CreateDepartureRequest request);
        DepartureDetailResponse AssignBookingsToDeparture(AssignBookingsToDepartureRequest request);
        DepartureDetailResponse? GetDepartureById(long departureId);
        List<DepartureListItem> GetAllDepartures();
    }

    // ================================================================
    // Service Implementation
    // ================================================================
    public class DepartureManagementService : IDepartureManagementService
    {
        private readonly IDepartureManagementRepository _repository;
        private readonly IMapper _mapper;
        private readonly INotificationService _notificationService;

        public DepartureManagementService(
            IDepartureManagementRepository repository,
            IMapper mapper,
            INotificationService notificationService)
        {
            _repository = repository;
            _mapper = mapper;
            _notificationService = notificationService;
        }

        // ─── Get all unassigned bookings grouped by itinerary ────────────
        public List<BookingGroupByItinerary> GetUnassignedBookingsGrouped()
        {
            var rows = _repository.GetUnassignedBookings();
            var domainRows = _mapper.Map<List<UnassignedBookingItem>>(rows);

            return domainRows
                .GroupBy(b => new { b.ItineraryId, b.ItineraryTitle })
                .Select(g => new BookingGroupByItinerary
                {
                    ItineraryId    = g.Key.ItineraryId,
                    ItineraryTitle = g.Key.ItineraryTitle,
                    TotalBookings  = g.Count(),
                    TotalPersons   = g.Sum(b => b.TotalPerson),
                    Bookings       = g.ToList()
                })
                .OrderBy(g => g.ItineraryTitle)
                .ToList();
        }

        // ─── Suggest departures with enough capacity ─────────────────────
        public List<SuggestedDepartureItem> GetSuggestedDepartures(long itineraryId, int requiredSeats)
        {
            var rows = _repository.GetSuggestedDepartures(itineraryId, requiredSeats);
            return _mapper.Map<List<SuggestedDepartureItem>>(rows);
        }

        // ─── Create new departure + assign bookings ───────────────────────
        public DepartureDetailResponse CreateDeparture(CreateDepartureRequest request)
        {
            var repoRequest = _mapper.Map<RepDTO.CreateDepartureRequest>(request);
            var repoResponse = _repository.CreateDeparture(repoRequest);
            var result = _mapper.Map<DepartureDetailResponse>(repoResponse);

            // Admin real-time notification
            _ = _notificationService.CreateAndBroadcastAsync(new Domain.Models.CreateNotification
            {
                Type    = "departure",
                Title   = "Departure Created",
                Message = $"Departure for '{result.ItineraryTitle}' on {result.DepartureDate:dd MMM yyyy} created with {result.BookedCount} traveler(s).",
                Link    = "/manage-bookings",
                Icon    = "pi-send"
            });

            // Notify each traveler
            NotifyTravelers(result.Id);

            return result;
        }

        // ─── Assign bookings to existing departure ────────────────────────
        public DepartureDetailResponse AssignBookingsToDeparture(AssignBookingsToDepartureRequest request)
        {
            var repoRequest = _mapper.Map<RepDTO.AssignBookingsToDepartureRequest>(request);
            var repoResponse = _repository.AssignBookingsToDeparture(repoRequest);
            var result = _mapper.Map<DepartureDetailResponse>(repoResponse);

            _ = _notificationService.CreateAndBroadcastAsync(new Domain.Models.CreateNotification
            {
                Type    = "departure",
                Title   = "Bookings Assigned to Departure",
                Message = $"{request.BookingIds.Count} booking(s) assigned to departure for '{result.ItineraryTitle}'.",
                Link    = "/manage-bookings",
                Icon    = "pi-users"
            });

            NotifyTravelers(result.Id);
            return result;
        }

        // ─── Get departure by id ──────────────────────────────────────────
        public DepartureDetailResponse? GetDepartureById(long departureId)
        {
            var repoResponse = _repository.GetDepartureById(departureId);
            return repoResponse == null ? null : _mapper.Map<DepartureDetailResponse>(repoResponse);
        }

        // ─── Get all departures with computed status ──────────────────────
        public List<DepartureListItem> GetAllDepartures()
        {
            var rows = _repository.GetAllDepartures();
            var today = DateOnly.FromDateTime(DateTime.UtcNow);

            return rows.Select(r =>
            {
                var item = _mapper.Map<DepartureListItem>(r);
                item.HasInventory = r.GuideId.HasValue || r.VehicleId.HasValue;

                item.ComputedStatus = r.Status == "Cancelled" ? "Cancelled"
                    : r.DepartureDate > today ? "Upcoming"
                    : r.DepartureDate == today ? "Ongoing"
                    : (r.EndDate.HasValue && r.EndDate.Value >= today) ? "Ongoing"
                    : "Completed";

                return item;
            }).ToList();
        }

        // ─── Internal: fire-and-forget traveler notifications ─────────────
        private void NotifyTravelers(long departureId)
        {
            try
            {
                var travelers = _repository.GetTravelerNotifications(departureId);
                foreach (var t in travelers)
                {
                    var endLabel = t.EndDate.HasValue ? t.EndDate.Value.ToString("dd MMM yyyy") : "TBD";
                    var msg = $"Your '{t.ItineraryTitle}' trip from {t.StartDate.ToString("dd MMM yyyy")} to {endLabel} is confirmed." +
                              (t.GuideName != null ? $" Guide: {t.GuideName}." : "") +
                              (t.VehicleInfo != null ? $" Vehicle: {t.VehicleInfo}." : "") +
                              $" Total group size: {t.TotalGroupSize} traveler(s).";

                    _ = _notificationService.CreateAndBroadcastAsync(new Domain.Models.CreateNotification
                    {
                        Type    = "trip-confirmed",
                        Title   = "Your Trip is Confirmed! ✈️",
                        Message = msg,
                        Link    = $"/booking-detail/{t.BookingInstanceId}",
                        Icon    = "pi-check-circle"
                    });
                }
            }
            catch
            {
                // Non-blocking: notification failures must not break the main flow
            }
        }
    }
}

File: frontend/src/app/pages/booking/departure-management.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

// ===========================
// INTERFACES
// ===========================

export interface UnassignedBookingItem {
    instanceId: number;
    bookingReference?: string;
    itineraryId: number;
    itineraryTitle: string;
    startDate?: string;
    endDate?: string;
    totalPerson: number;
    status: string;
    paymentStatus: string;
    totalAmount: number;
    createdAt: string;
    leadTravelerName?: string;
    leadTravelerEmail?: string;
    leadTravelerPhone?: string;
}

export interface BookingGroupByItinerary {
    itineraryId: number;
    itineraryTitle: string;
    totalBookings: number;
    totalPersons: number;
    bookings: UnassignedBookingItem[];
}

/** Sub-group within an itinerary, bookings sharing the same start date */
export interface DateSubGroup {
    startDate: string;          // ISO date string or 'unset'
    bookings: UnassignedBookingItem[];
    totalPersons: number;
}

export interface SuggestedDepartureItem {
    id: number;
    itineraryTitle: string;
    departureDate: string;
    endDate?: string;
    availableSeats: number;
    bookedCount: number;
    capacity: number;
    status: string;
    guideName?: string;
    vehicleInfo?: string;
}

export interface CreateDepartureRequest {
    itineraryId: number;
    startDate: string;
    endDate: string;
    capacity: number;
    guideId?: number;
    vehicleId?: number;
    bookingIds: number[];
    packagePrice: number;
    notes?: string;
}

export interface AssignBookingsToDepartureRequest {
    departureId: number;
    bookingIds: number[];
}

export interface DepartureDetailResponse {
    id: number;
    itineraryId: number;
    itineraryTitle: string;
    departureDate: string;
    endDate?: string;
    capacity: number;
    bookedCount: number;
    availableSeats: number;
    status: string;
    packagePrice: number;
    guideId?: number;
    guideName?: string;
    vehicleId?: number;
    vehicleInfo?: string;
    notes?: string;
    createdAt: string;
    assignedBookings: UnassignedBookingItem[];
}

export interface DepartureListItem {
    id: number;
    itineraryId: number;
    itineraryTitle: string;
    departureDate: string;
    endDate?: string;
    capacity: number;
    bookedCount: number;
    availableSeats: number;
    status: string;
    computedStatus: string;   // Upcoming | Ongoing | Completed | Cancelled
    packagePrice: number;
    guideId?: number;
    guideName?: string;
    guideSpecialization?: string;
    vehicleId?: number;
    vehicleInfo?: string;
    hasInventory: boolean;
    travelerCount: number;
    travelerNames?: string;
    notes?: string;
    createdAt: string;
}

// ===========================
// SERVICE
// ===========================
@Injectable({ providedIn: 'root' })
export class DepartureManagementService {
    private baseUrl = `${environment.apiBaseUrl}/departure-management`;

    constructor(private http: HttpClient) {}

    getUnassignedBookings(): Observable<BookingGroupByItinerary[]> {
        return this.http.get<BookingGroupByItinerary[]>(`${this.baseUrl}/unassigned-bookings`);
    }

    getSuggestedDepartures(itineraryId: number, requiredSeats: number): Observable<SuggestedDepartureItem[]> {
        const params = new HttpParams()
            .set('itineraryId', itineraryId)
            .set('requiredSeats', requiredSeats);
        return this.http.get<SuggestedDepartureItem[]>(`${this.baseUrl}/suggest`, { params });
    }

    createDeparture(request: CreateDepartureRequest): Observable<DepartureDetailResponse> {
        return this.http.post<DepartureDetailResponse>(`${this.baseUrl}/create`, request);
    }

    assignBookings(request: AssignBookingsToDepartureRequest): Observable<DepartureDetailResponse> {
        return this.http.post<DepartureDetailResponse>(`${this.baseUrl}/assign`, request);
    }

    getDepartureById(id: number): Observable<DepartureDetailResponse> {
        return this.http.get<DepartureDetailResponse>(`${this.baseUrl}/${id}`);
    }

    getAllDepartures(): Observable<DepartureListItem[]> {
        return this.http.get<DepartureListItem[]>(`${this.baseUrl}/all`);
    }
}
